src/main-process/atom-application.js

```diff
@@ -278,23 +268,18 @@ module.exports = class AtomApplication extends EventEmitter {
       this.safeMode
     );
 
-    // Don't await for the following method to avoid delaying the opening of a new window.
-    // (we await it just after opening it).
-    // We need to do this because `listenForArgumentsFromNewProcess()` calls `crypto.randomBytes`,
-    // which is really slow on Windows machines.
-    // (TodoElectronIssue: This got fixed in electron v3: https://github.com/electron/electron/issues/2073).
     let socketServerPromise;
     if (options.test || options.benchmark || options.benchmarkTest) {
       socketServerPromise = Promise.resolve();
     } else {
       socketServerPromise = this.listenForArgumentsFromNewProcess();
     }
 
+    await socketServerPromise;
```

Comment:
What is the benefit of awaiting this earlier? If it is not needed why we should do it?

Reply:
@DeeDeeG

Reply:
> Benchmarking the await would help as well.

If someone other than me could handle this, that would be great. I'm all for learning new things, but someone who has benchmarked JS before might better be able to tell whether the results look good or bad.

> What is the benefit of awaiting this earlier?

Supposedly it prevents a race condition. This listener listens for requests to open new windows, and somehow transfers these requests/their metadata to the main window or process. (According to the PRs I link below). And it really isn't much slower, maybe under ~20ms usually? Or not slower at all? Hard to tell, honestly, since startups are so variable already. But I used the builtin [`timecop`](https://github.com/atom/timecop) package, and it's similar with or without these tweaks. I am relying on the commentary of Atom devs in the pull requests that implemented the workaround in the first place. See discussion here: https://github.com/atom/atom/pull/19242 & https://github.com/atom/atom/pull/19246 -- this code was tweaked a bit more in https://github.com/atom/atom/pull/19354 & https://github.com/atom/atom/pull/19455.

Reply:
There is a race condition regardless. If you start two instances concurrently with `--inspect-brk=<port>` to pause them you can walk through and have both think they are establishing the socket. Both will check for the socket secret file (determined by Atom home + version; but none exists), both will generate independent secrets (just a random value), both will write their secret to the same socket secret file (only the last to write will persist), and both will establish servers listening to the named pipe corresponding to their secret.

An instance has already decided to keep going once this point is reached, and no socket related behaviour needs to happen strictly after the socket is established, so in terms of correctness it does not matter where the socket promise is awaited. AFAICT, the only reason to await it at all is to measure it's impact. The promise it becomes a part of is [not used](https://github.com/atom/atom/blob/72709dcaa5a6caf4fb7ea5426ffbaceba6ea1083/src/main-process/start.js#L124). 

(the use of the socket secret file is for new instances to find it, read the contained secret, deterministically convert this secret into the name of a named pipe for inter process communication, and inform the existing instance monitoring this pipe of the arguments to this new one).

Reply:
Thanks for that info.

With the debugger off, I think there is some (small) real-world impact. When I tried spamming "atom" in `cmd.exe` many times in a second, with Atom not started up yet, I informally was able to reproduce that the Atom windows ended up in fewer independent processes with this change. Would have to try again and confirm before I was totally confident in this result being meaningful though.

I think the impact is more profound with a very slow/resources-tied-up machine that really chugs during startup. (And maybe moreso for "cold starts"? Atom is dramatically slower to start when it hasn't been cached in RAM yet, in my experience.)